File: frontend/src/lib/api/guardrails.ts
/**
 * Guardrails API client for AI content filtering.
 *
 * Provides methods for managing guardrails at org, team, and user levels.
 */

import { apiClient, getAuthHeader } from "./client";

/** Action to take when a guardrail matches */
export type GuardrailAction = "block" | "warn" | "redact";

/** Available PII types for detection */
export const PII_TYPES = [
  "email",
  "phone",
  "ssn",
  "credit_card",
  "ip_address",
] as const;

export type PIIType = (typeof PII_TYPES)[number];

/** PII type display labels */
export const PII_TYPE_LABELS: Record<PIIType, string> = {
  email: "Email Addresses",
  phone: "Phone Numbers",
  ssn: "Social Security Numbers",
  credit_card: "Credit Card Numbers",
  ip_address: "IP Addresses",
};

/** Base guardrail settings shared across all levels */
export interface GuardrailSettingsBase {
  guardrails_enabled: boolean;
  input_blocked_keywords: string[];
  input_blocked_patterns: string[];
  input_action: GuardrailAction;
  output_blocked_keywords: string[];
  output_blocked_patterns: string[];
  output_action: GuardrailAction;
  pii_detection_enabled: boolean;
  pii_types: PIIType[];
  pii_action: GuardrailAction;
}

/** Organization guardrails with additional controls */
export interface OrganizationGuardrails extends GuardrailSettingsBase {
  id: string;
  organization_id: string;
  allow_team_override: boolean;
  allow_user_override: boolean;
  created_at: string;
  updated_at: string;
}

/** Team guardrails */
export interface TeamGuardrails extends GuardrailSettingsBase {
  id: string;
  team_id: string;
  created_at: string;
  updated_at: string;
}

/** User guardrails */
export interface UserGuardrails extends GuardrailSettingsBase {
  id: string;
  user_id: string;
  created_at: string;
  updated_at: string;
}

/** Update schema for organization guardrails */
export interface OrganizationGuardrailsUpdate {
  guardrails_enabled?: boolean;
  input_blocked_keywords?: string[];
  input_blocked_patterns?: string[];
  input_action?: GuardrailAction;
  output_blocked_keywords?: string[];
  output_blocked_patterns?: string[];
  output_action?: GuardrailAction;
  pii_detection_enabled?: boolean;
  pii_types?: PIIType[];
  pii_action?: GuardrailAction;
  allow_team_override?: boolean;
  allow_user_override?: boolean;
}

/** Update schema for team guardrails */
export interface TeamGuardrailsUpdate {
  guardrails_enabled?: boolean;
  input_blocked_keywords?: string[];
  input_blocked_patterns?: string[];
  input_action?: GuardrailAction;
  output_blocked_keywords?: string[];
  output_blocked_patterns?: string[];
  output_action?: GuardrailAction;
  pii_detection_enabled?: boolean;
  pii_types?: PIIType[];
  pii_action?: GuardrailAction;
}

/** Update schema for user guardrails */
export interface UserGuardrailsUpdate {
  guardrails_enabled?: boolean;
  input_blocked_keywords?: string[];
  input_blocked_patterns?: string[];
  input_action?: GuardrailAction;
  output_blocked_keywords?: string[];
  output_blocked_patterns?: string[];
  output_action?: GuardrailAction;
  pii_detection_enabled?: boolean;
  pii_types?: PIIType[];
  pii_action?: GuardrailAction;
}

/** Effective guardrails after applying hierarchy */
export interface EffectiveGuardrails {
  guardrails_enabled: boolean;
  guardrails_disabled_by: "org" | "team" | null;
  input_blocked_keywords: string[];
  input_blocked_patterns: string[];
  input_action: GuardrailAction;
  output_blocked_keywords: string[];
  output_blocked_patterns: string[];
  output_action: GuardrailAction;
  pii_detection_enabled: boolean;
  pii_types: PIIType[];
  pii_action: GuardrailAction;
  can_user_modify: boolean;
}

/** Match information from guardrail test */
export interface GuardrailMatch {
  pattern: string;
  pattern_type: "keyword" | "regex" | "pii";
  matched_text: string;
  start: number;
  end: number;
}

/** Result of testing content against guardrails */
export interface GuardrailTestResult {
  passed: boolean;
  action: GuardrailAction | null;
  matches: GuardrailMatch[];
  redacted_content: string | null;
}

/** Request for testing guardrails */
export interface GuardrailTestRequest {
  content: string;
  direction: "input" | "output";
}

export const guardrailsApi = {
  /**
   * Get organization guardrails
   */
  async getOrgGuardrails(orgId: string): Promise<OrganizationGuardrails> {
    return apiClient.get<OrganizationGuardrails>(
      `/v1/guardrails/organizations/${orgId}`,
      { headers: getAuthHeader() },
    );
  },

  /**
   * Update organization guardrails
   */
  async updateOrgGuardrails(
    orgId: string,
    data: OrganizationGuardrailsUpdate,
  ): Promise<OrganizationGuardrails> {
    return apiClient.put<OrganizationGuardrails>(
      `/v1/guardrails/organizations/${orgId}`,
      data,
      { headers: getAuthHeader() },
    );
  },

  /**
   * Get team guardrails
   */
  async getTeamGuardrails(
    orgId: string,
    teamId: string,
  ): Promise<TeamGuardrails> {
    return apiClient.get<TeamGuardrails>(
      `/v1/guardrails/organizations/${orgId}/teams/${teamId}`,
      { headers: getAuthHeader() },
    );
  },

  /**
   * Update team guardrails
   */
  async updateTeamGuardrails(
    orgId: string,
    teamId: string,
    data: TeamGuardrailsUpdate,
  ): Promise<TeamGuardrails> {
    return apiClient.put<TeamGuardrails>(
      `/v1/guardrails/organizations/${orgId}/teams/${teamId}`,
      data,
      { headers: getAuthHeader() },
    );
  },

  /**
   * Get user guardrails
   */
  async getUserGuardrails(): Promise<UserGuardrails> {
    return apiClient.get<UserGuardrails>("/v1/guardrails/me", {
      headers: getAuthHeader(),
    });
  },

  /**
   * Update user guardrails
   */
  async updateUserGuardrails(
    data: UserGuardrailsUpdate,
  ): Promise<UserGuardrails> {
    return apiClient.put<UserGuardrails>("/v1/guardrails/me", data, {
      headers: getAuthHeader(),
    });
  },

  /**
   * Get effective guardrails (computed from hierarchy)
   */
  async getEffectiveGuardrails(
    orgId?: string,
    teamId?: string,
  ): Promise<EffectiveGuardrails> {
    const params = new URLSearchParams();
    if (orgId) params.set("org_id", orgId);
    if (teamId) params.set("team_id", teamId);

    return apiClient.get<EffectiveGuardrails>(
      `/v1/guardrails/effective?${params.toString()}`,
      { headers: getAuthHeader() },
    );
  },

  /**
   * Test content against guardrails (dry run)
   */
  async testGuardrails(
    request: GuardrailTestRequest,
    orgId?: string,
    teamId?: string,
  ): Promise<GuardrailTestResult> {
    const params = new URLSearchParams();
    if (orgId) params.set("org_id", orgId);
    if (teamId) params.set("team_id", teamId);

    return apiClient.post<GuardrailTestResult>(
      `/v1/guardrails/test?${params.toString()}`,
      request,
      { headers: getAuthHeader() },
    );
  },

  /**
   * Get available PII types
   */
  async getPIITypes(): Promise<PIIType[]> {
    return apiClient.get<PIIType[]>("/v1/guardrails/pii-types", {
      headers: getAuthHeader(),
    });
  },
};

File: backend/src/backend/api/routes/guardrails.py
"""Guardrails API routes for AI content filtering.

Provides endpoints for managing guardrails at org, team, and user levels.
"""

from typing import Annotated
import uuid

from fastapi import APIRouter, Depends, HTTPException, Query

from backend.auth import CurrentUser, SessionDep
from backend.guardrails.models import (
    EffectiveGuardrails,
    GuardrailsTestRequest,
    GuardrailsTestResponse,
    OrganizationGuardrailsPublic,
    OrganizationGuardrailsUpdate,
    TeamGuardrailsPublic,
    TeamGuardrailsUpdate,
    UserGuardrailsPublic,
    UserGuardrailsUpdate,
)
from backend.guardrails.patterns import PII_TYPES
from backend.guardrails.service import (
    get_effective_guardrails,
    get_or_create_org_guardrails,
    get_or_create_team_guardrails,
    get_or_create_user_guardrails,
    test_guardrails,
    update_org_guardrails,
    update_team_guardrails,
    update_user_guardrails,
)
from backend.rbac import (
    OrgContextDep,
    OrgPermission,
    RequireOrgAdminDep,
    RequireTeamAdminDep,
    TeamContextDep,
    require_org_permission,
)

router = APIRouter(prefix="/guardrails", tags=["guardrails"])


# Organization guardrails routes
@router.get(
    "/organizations/{organization_id}",
    response_model=OrganizationGuardrailsPublic,
    dependencies=[Depends(require_org_permission(OrgPermission.ORG_READ))],
)
async def get_org_guardrails(
    session: SessionDep,
    org_context: OrgContextDep,
    _admin: RequireOrgAdminDep,
) -> OrganizationGuardrailsPublic:
    """Get guardrails for an organization.

    Requires org admin or owner role.
    """
    guardrails = get_or_create_org_guardrails(session, org_context.organization.id)
    return OrganizationGuardrailsPublic.model_validate(guardrails)


@router.put(
    "/organizations/{organization_id}",
    response_model=OrganizationGuardrailsPublic,
    dependencies=[Depends(require_org_permission(OrgPermission.ORG_UPDATE))],
)
async def update_org_guardrails_endpoint(
    data: OrganizationGuardrailsUpdate,
    session: SessionDep,
    org_context: OrgContextDep,
    _admin: RequireOrgAdminDep,
) -> OrganizationGuardrailsPublic:
    """Update guardrails for an organization.

    Requires org admin or owner role.
    """
    guardrails = update_org_guardrails(session, org_context.organization.id, data)
    return OrganizationGuardrailsPublic.model_validate(guardrails)


# Team guardrails routes
@router.get(
    "/organizations/{organization_id}/teams/{team_id}",
    response_model=TeamGuardrailsPublic,
)
async def get_team_guardrails(
    session: SessionDep,
    team_context: TeamContextDep,
    _admin: RequireTeamAdminDep,
) -> TeamGuardrailsPublic:
    """Get guardrails for a team.

    Requires team admin role.
    """
    # Check if org allows team override
    org_guardrails = get_or_create_org_guardrails(
        session, team_context.org_context.organization.id
    )
    if not org_guardrails.allow_team_override:
        raise HTTPException(
            status_code=403,
            detail="Organization does not allow team guardrail customization",
        )

    guardrails = get_or_create_team_guardrails(session, team_context.team.id)
    return TeamGuardrailsPublic.model_validate(guardrails)


@router.put(
    "/organizations/{organization_id}/teams/{team_id}",
    response_model=TeamGuardrailsPublic,
)
async def update_team_guardrails_endpoint(
    data: TeamGuardrailsUpdate,
    session: SessionDep,
    team_context: TeamContextDep,
    _admin: RequireTeamAdminDep,
) -> TeamGuardrailsPublic:
    """Update guardrails for a team.

    Requires team admin role.
    """
    # Check if org allows team override
    org_guardrails = get_or_create_org_guardrails(
        session, team_context.org_context.organization.id
    )
    if not org_guardrails.allow_team_override:
        raise HTTPException(
            status_code=403,
            detail="Organization does not allow team guardrail customization",
        )

    guardrails = update_team_guardrails(session, team_context.team.id, data)
    return TeamGuardrailsPublic.model_validate(guardrails)


# User guardrails routes
@router.get(
    "/me",
    response_model=UserGuardrailsPublic,
)
async def get_user_guardrails(
    session: SessionDep,
    current_user: CurrentUser,
) -> UserGuardrailsPublic:
    """Get guardrails for the current user."""
    guardrails = get_or_create_user_guardrails(session, current_user.id)
    return UserGuardrailsPublic.model_validate(guardrails)


@router.put(
    "/me",
    response_model=UserGuardrailsPublic,
)
async def update_user_guardrails_endpoint(
    data: UserGuardrailsUpdate,
    session: SessionDep,
    current_user: CurrentUser,
) -> UserGuardrailsPublic:
    """Update guardrails for the current user."""
    guardrails = update_user_guardrails(session, current_user.id, data)
    return UserGuardrailsPublic.model_validate(guardrails)


# Effective guardrails route
@router.get(
    "/effective",
    response_model=EffectiveGuardrails,
)
async def get_effective_guardrails_endpoint(
    session: SessionDep,
    current_user: CurrentUser,
    org_id: Annotated[
        str | None, Query(description="Organization ID for context")
    ] = None,
    team_id: Annotated[str | None, Query(description="Team ID for context")] = None,
) -> EffectiveGuardrails:
    """Get effective guardrails after applying hierarchy.

    This computes the merged guardrails from org -> team -> user levels.
    """
    org_uuid = uuid.UUID(org_id) if org_id else None
    team_uuid = uuid.UUID(team_id) if team_id else None

    return get_effective_guardrails(
        session=session,
        user_id=current_user.id,
        organization_id=org_uuid,
        team_id=team_uuid,
    )


# Test guardrails route
@router.post(
    "/test",
    response_model=GuardrailsTestResponse,
)
async def test_guardrails_endpoint(
    request: GuardrailsTestRequest,
    session: SessionDep,
    current_user: CurrentUser,
    org_id: Annotated[
        str | None, Query(description="Organization ID for context")
    ] = None,
    team_id: Annotated[str | None, Query(description="Team ID for context")] = None,
) -> GuardrailsTestResponse:
    """Test content against guardrails (dry run).

    This allows users to test their guardrail configuration without
    actually blocking content.
    """
    org_uuid = uuid.UUID(org_id) if org_id else None
    team_uuid = uuid.UUID(team_id) if team_id else None

    effective = get_effective_guardrails(
        session=session,
        user_id=current_user.id,
        organization_id=org_uuid,
        team_id=team_uuid,
    )

    result = test_guardrails(
        content=request.content,
        direction=request.direction,
        guardrails=effective,
    )

    return GuardrailsTestResponse(
        passed=result.passed,
        action=result.action,
        matches=result.matches,
        redacted_content=result.redacted_content,
    )


# PII types reference
@router.get(
    "/pii-types",
    response_model=list[str],
)
async def get_pii_types() -> list[str]:
    """Get available PII types for detection."""
    return PII_TYPES
